### PR TITLE
align metric queries by step and other queries by split interval

### PR DESCRIPTION
**Pull request by sandeepsukhani**
**What this PR does / why we need it**:
It changes the way query splits are built based on the query type:
1. Metric queries: Start/End time in query splits would be aligned by `step` in query.
2. All other queries: Start/End time in query splits would be aligned by the split interval.

*Note:* Start of the first query and end time of the last query is kept the same as the original query.

This change is being done for two reasons:
1. The step alignment code(when enabled) rounds down both Start and end time which could modify it too much if the steps are too large.
2. Metadata queries perform well when aligned by `24h`(split interval). Testing this internally reduced 3x the number of index queries and improved query performance by 2x for long-range `/label` calls.

While there is no visible impact on the results cache hit rate, I think we should also change the step alignment code to round down the start time and round up the end time, which would anyways only align the start time of the first query and end time of the last query.
image

**Checklist**
- [x] Tests updated

### CODE REVIEW

File: pkg/querier/queryrange/split_by_interval.go [hunk 1]
```diff
@@ -253,11 +253,26 @@ func splitByTime(req queryrangebase.Request, interval time.Duration) []queryrang
...
+	firstInterval := true
+
 	for start := start; start.Before(end); start = start.Add(interval) {
 		newEnd := start.Add(interval)
-		if newEnd.After(end) {
+		if !newEnd.Before(end) {
 			newEnd = end
+		} else if endTimeInclusive {
+			newEnd = newEnd.Add(-time.Millisecond)
+		}
```
Comment (cyriltovena) [inline pkg/querier/queryrange/split_by_interval.go:270]: I don't quite understand this.

File: pkg/querier/queryrange/split_by_interval.go [hunk 2]
```diff
@@ -298,6 +318,9 @@ func splitMetricByTime(r queryrangebase.Request, interval time.Duration) []query
 			EndTs:     end,
 		})
 	}
+
+	// change the start time to original time
+	reqs[0] = reqs[0].WithStartEnd(lokiReq.GetStart(), reqs[0].GetEnd())
```
Comment (cyriltovena) [inline pkg/querier/queryrange/split_by_interval.go:323]: can you add a check `if len(reqs) != 0`

**Comment by cyriltovena** [comment]
Should we deprecate the `StepAlign` middleware ?

File: pkg/querier/queryrange/split_by_interval.go (hunk 1 above)
Comment (sandeepsukhani) [reply pkg/querier/queryrange/split_by_interval.go:270]: When the end time is inclusive, we do not want the current query's end time to be the same as the start time of the next query. The only queries that have both start and end time inclusive are metadata queries, and without this change, we would still end up querying the same data as the next query.
I will add a comment to make it clear.

**Review by cyriltovena — APPROVED** [review]
LGTM

**Comment by sandeepsukhani** [comment]
> Should we deprecate the `StepAlign` middleware ?

While there is not much value in keeping it, I was thinking about changing it to round down the start time and round up the end time to make all the queries cacheable.

**Pull request merged**